### PR TITLE
:tada: store TODO on redux by saga

### PR DESCRIPTION
**Pull request by guillaumecrespel**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -0,0 +1,157 @@
+{
+  "name": "trampss-todomvc",
+  "version": "0.1.0",
+  "description": "TodoMVC with React / Redux / Redux-Saga",
+  "license": "MIT",
+  "scripts": {
+    "start": "cross-env NODE_ENV=development webpack-dev-server --host 0.0.0.0 --inline --hot --port=3000 --history-api-fallback --watch-poll=1000",
+    "build": "cross-env NODE_ENV=production webpack -p",
+    "lint": "eslint --cache --ext js,jsx src",
```
Comment (fabienjuif) [inline package.json]: Look at this : https://github.com/Zenika/react-workbench/blob/master/package.json#L13

File: package.json [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+  "license": "MIT",
+  "scripts": {
+    "start": "cross-env NODE_ENV=development webpack-dev-server --host 0.0.0.0 --inline --hot --port=3000 --history-api-fallback --watch-poll=1000",
+    "build": "cross-env NODE_ENV=production webpack -p",
+    "lint": "eslint --cache --ext js,jsx src",
+    "test": "jest",
+    "analyze": "webpack --json > bundle.analyze.json"
+  },
+  "dependencies": {
+    "babel-plugin-transform-inline-environment-variables": "^0.0.2",
```
Comment (fabienjuif) [inline package.json]: - this is a dev dependency
 - I think we don't need it in this project

File: package.json [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    "babel-polyfill": "^6.20.0",
+    "hoc-little-router": "^1.0.0",
```
Comment (fabienjuif) [inline package.json]: TodoMVC has routes to handle ?

File: package.json [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    "isomorphic-fetch": "2.2.1",
+    "lodash": "^4.17.4",
```
Comment (fabienjuif) [inline package.json]: - needed ?

File: package.json [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    "react-redux": "^5.0.2",
+    "recompose": "^0.22.0",
+    "redux": "^3.6.0",
+    "redux-little-router": "^13.1.2",
+    "redux-saga": "^0.14.3",
+    "todomvc-app-css": "^2.1.0",
+    "trampss-redux-factory": "^3.5.0",
+    "trampss-redux-saga-tester": "^1.0.0"
+  },
+  "devDependencies": {
```
Comment (fabienjuif) [inline package.json:31]: I have the feeling that dependencies can be stripped out : 
 - remove all dependencies
Then : 
 - run test / build / lint 
 - add missing dependency
 - loop

File: src/redux/reducers.js [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import { combineReducers } from 'redux'
+import { routerForBrowser } from 'redux-little-router'
+import router from './router'
+import ui from './ui'
+import todos from './todos'
+
+export const CONTEXT = process.env.NODE_ENV === 'development' ? '' : '/todomvc'
```
Comment (fabienjuif) [inline src/redux/reducers.js]: not needed here

File: src/redux/todos.js [hunk 7]
```diff
@@ -0,0 +1,6 @@
+import { keyValue } from 'trampss-redux-factory'
+import { mapPayload } from 'trampss-redux-factory/helpers'
+
+const mapper = todo => ({ id: new Date().getTime(), todo })
+
+export default keyValue({ pre: [mapPayload(/ADD_TODOS/)(mapper)] })({ key: 'id', name: 'todos' })
```
Comment (fabienjuif) [inline src/redux/todos.js]: Something like `/@trampss\/ADD_TODOS/` ?

File: src/redux/reducers.js [hunk 8]
```diff
@@ -0,0 +1,15 @@
...
+
+const { reducer } = routerForBrowser({ routes: router, basename: CONTEXT })
+
+export default combineReducers({
+  router: reducer,
+  todos,
```
Comment (fabienjuif) [inline src/redux/reducers.js]: I think our pattern is good and we should commit into it : 
 - `ui`
    - router
    - form
    - etc
 - `data` (previously `api`)
    - todos

File: src/components/header/header.jsx [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+const Header = ({ newTodo, addTodo, setNewTodo }) => {
+  return (
+    <header className="header">
+      <h1>todos</h1>
+      <input
+        className="new-todo"
+        placeholder="What needs to be done?"
+        value={newTodo}
+        onKeyDown={addTodo}
+        onChange={setNewTodo}
```
Comment (fabienjuif) [inline src/components/header/header.jsx]: onTodoChange

File: src/components/header/header.actions.js [hunk 10]
```diff
@@ -0,0 +1,2 @@
+export const addTodo = todo => ({ type: 'ADD_NEW_TODO', payload: todo })
+export const setNewTodo = todo => ({ type: 'SET_NEW_TODO', payload: todo })
```
Comment (fabienjuif) [inline src/components/header/header.actions.js]: `CHANGE_NEW_TODO` or `CHANGE_TODO` ?

File: src/components/header/header.container.js [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+export const mapStateToProps = (state) => {
+  return {
+    newTodo: newTodo.get()(state),
+  }
+}
+
+export const mapDispatchToProps = (dispatch) => {
+  return {
+    addTodo: e => checkEnter(e.keyCode) && dispatch(addTodo(e.target.value)),
+    setNewTodo: e => dispatch(setNewTodo(e.target.value)),
```
Comment (fabienjuif) [inline src/components/header/header.container.js]: onTodoChange

File: src/components/header/header.actions.js (hunk 10 above)
Comment (fabienjuif) [reply src/components/header/header.actions.js]: What about using redux-form ?

**Review by fabienjuif — CHANGES_REQUESTED** [review]
Look at comments

File: package.json (hunk 1 above)
Comment (guillaumecrespel) [reply package.json]: 👌

File: package.json (hunk 2 above)
Comment (guillaumecrespel) [reply package.json]: 👌

File: package.json (hunk 3 above)
Comment (guillaumecrespel) [reply package.json]: I think is good to show hoc-little-router, i find an utility ;)

File: package.json (hunk 4 above)
Comment (guillaumecrespel) [reply package.json]: 👌

File: src/redux/reducers.js (hunk 6 above)
Comment (guillaumecrespel) [reply src/redux/reducers.js]: 👌

File: src/redux/todos.js (hunk 7 above)
Comment (guillaumecrespel) [reply src/redux/todos.js]: 👌

File: src/redux/reducers.js (hunk 8 above)
Comment (guillaumecrespel) [reply src/redux/reducers.js]: 👌

File: package.json [hunk 12]
```diff
@@ -0,0 +1,153 @@
...
+  "scripts": {
+    "start": "webpack-dev-server --host 0.0.0.0 --inline --hot --port=3000 --history-api-fallback --watch-poll=1000",
+    "lint:js": "eslint --ext js,jsx src",
+    "lint": "npm-run-all --parallel lint:*",
+    "test": "jest",
+    "analyze": "webpack --json > bundle.analyze.json"
+  },
+  "dependencies": {
+    "babel-polyfill": "^6.20.0",
+    "isomorphic-fetch": "2.2.1",
```
Comment (fabienjuif) [inline package.json]: Has TodoMVC some API fetch ?
If not, you could remove this dependency

File: src/redux/data/index.js [hunk 13]
```diff
@@ -0,0 +1,10 @@
+import { combineReducers } from 'redux'
+import todos from './todos'
+
+/* Reducer name */
+export const path = 'data'
```
Comment (fabienjuif) [inline src/redux/data/index.js]: Unused variable ?

File: src/redux/data/todos.js [hunk 14]
```diff
@@ -0,0 +1,6 @@
+import { keyValue } from 'trampss-redux-factory'
+import { mapPayload } from 'trampss-redux-factory/helpers'
+
+/* Todos Reducer  with her mapper */
+const todosMapper = todo => ({ id: new Date().getTime(), todo })
+export default keyValue({ pre: [mapPayload(/@trampss\/ADD_TODOS/)(todosMapper)] })({ path: 'data', key: 'id', name: 'todos' })
```
Comment (fabienjuif) [inline src/redux/data/todos.js:6]: You don't use the exported path from data.js

File: src/components/header/header.actions.js (hunk 10 above)
Comment (guillaumecrespel) [reply src/components/header/header.actions.js]: - CHANGE_TODO its for edit an existing TODO, and NEW_TODO its the name of input.
- I find that the goal is to show our libraries and redux-form adds yet a level of abstraction and makes it more difficult to understand

File: src/components/header/header.container.js (hunk 11 above)
Comment (guillaumecrespel) [reply src/components/header/header.container.js]: 👌

File: package.json (hunk 12 above)
Comment (guillaumecrespel) [reply package.json]: i have remove this

File: src/redux/data/todos.js (hunk 14 above)
Comment (guillaumecrespel) [reply src/redux/data/todos.js:6]: 👍

File: package.json (hunk 3 above)
Comment (guillaumecrespel) [reply package.json]: i finded ! it's for filter on initial project !

**Pull request merged**